src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs

```diff
@@ -210,15 +210,24 @@ private enum Flags : byte
                                               hasInitializer: (_propertyFlags & Flags.HasInitializer) != 0,
                                               isCreatedForFieldKeyword: isCreatedForFieldKeyword,
                                               isEarlyConstructed: isEarlyConstructed);
-                Interlocked.CompareExchange(ref _lazyBackingFieldSymbol, backingField, _lazyBackingFieldSymbolSentinel);
+                if (Interlocked.CompareExchange(ref _lazyBackingFieldSymbol, backingField, _lazyBackingFieldSymbolSentinel) == _lazyBackingFieldSymbolSentinel &&
+                    isCreatedForFieldKeyword)
+                {
+                    // semi auto property should override all accessors.
+                    if ((!HasSetAccessor && !this.IsReadOnly) ||
+                        (!HasGetAccessor && !this.IsWriteOnly))
+                    {
+                        diagnostics.Add(ErrorCode.ERR_AutoPropertyMustOverrideSet, Location, this);
```

Comment:
We should report errors in `SourcePropertySymbolBase.AfterAddingTypeMembersChecks`, not as part of binding `field` to a value. That should remove the need to pass around a diagnostics bag as part of binding the field symbol.

Reply:
Accessing FieldKeywordBackingField in AfterAddingTypeMembersChecks (before binding) will cause extra binding, which @AlekseyTs wanted to avoid whenever possible.

Reply:
Then we should find a different location to insert the check. I don't think the check is appropriate in the current location.

Reply:
@333fred Will doing that like `CircularStructDiagnostics` in #60601 be good?

Reply:
> I'd like to know why and what problems can this cause :)

This section of binding is about binding an individual member. It should be logically separated from things like checking that overrides are performed correctly, as "you must override both get and set" is unrelated binding the body of either get or set.

As to the appropriate location, I think a better approach would be a delayed diagnostic. We can take an approach similar to how `LazyObsoleteDiagnosticInfo` works: during `AfterAddingTypeMembersChecks`, we check to see whether the backing field has been calculated. If it has, we can error right away. If not, we add a lazy diagnostic. When the diagnostic is pulled, we ensure that the backing field is calculated (forcing binding as necessary), and then perform the check.

Reply:
But I will also let @AlekseyTs chime in if he has other ideas.

Reply:
Could we use `CompletionPart` for it? From its documentation comment:

```csharp
    /// <summary>
    /// This enum describes the types of components that could give
    /// us diagnostics.  We shouldn't read the list of diagnostics
    /// until all of these types are accounted for.
    /// </summary>
```

That *looks* suitable for this scenario?

Reply:
I'd be ok with the `MarkBackingFieldsAsCalculated` approach.

Reply:
@AlekseyTs @333fred Are we sure we'll always call `MarkBackingFieldAsCalculated`? (I think no)

If yes, are we sure it's going to be called only once? Otherwise, I'd need some sort of lock to prevent races causing duplicate diagnostics.

---

> to report this error into MethodCompiler's bag.

MethodCompiler isn't the only caller. Is there a problem if we always report the diagnostic into the bag given to the `SourcePropertySymbolBase` constructor?

---

> Alternatively, if we don't mind reporting this error per field usage, we could report this check in the binder, where ERR_InterfacesCantContainFields is reported.

I'm more with this approach, but instead of reporting on every field usage. We report once and use the property location, not `field` location.

And even then, we'll have a wrong behavior for initializers where we early create the backing field. This will another check and tests.

@333fred @AlekseyTs What do you think?

Reply:
@AlekseyTs A bad commit: https://github.com/Youssef1313/roslyn/commit/b714f01b0abcd271c6ee0ec1f10929058262c86d

It's very flaky. Not only the new test. But I see other existing tests can randomly have the diagnostics from the compilation of the new test. I can't understand how the diagnostics leaked into a whole different compilation.

Reply:
> Alternatively, if we don't mind reporting this error per field usage, we could report this check in the binder, where ERR_InterfacesCantContainFields is reported.

I'm using this approach now.

Reply:
I really do not like this approach, as it results in duplicated diagnostics. I think a delayed diagnostic will address the concerns and ensure we don't have duplicate diagnostics.

Reply:
Duplicate diagnostics can be fixed easily I think. The other approach didn't work well with me (see my previous comment). If you can help getting things stable with the other approach, I don't mind using it.

Reply:
> I think a delayed diagnostic will address the concerns and ensure we don't have duplicate diagnostics.

I don't think the delayed diagnostics is the right approach here. Mainly because we will have to control when it is getting resolved, which is hard to do given that we are trying to move the work from one pipeline phase to the other.

>It's very flaky. Not only the new test. But I see other existing tests can randomly have the diagnostics from the compilation of the new test. I can't understand how the diagnostics leaked into a whole different compilation.

>It's likely I'm reusing a diagnostic bag that shouldn't be reused, but still looks very weird it's leaking into a different compilation.

Let's take a look at what you tried to do and see what is going wrong. Update this PR with that. I have a high confidence that ```MarkBackingFieldAsCalculated``` approach will work. In fact, it is very likely that later we will move/add reporting of even more diagnostics there.

Reply:
> Duplicate diagnostics can be fixed easily I think. The other approach didn't work well with me (see my previous comment). If you can help getting things stable with the other approach, I don't mind using it. 

Please update the PR with the approach so that we could comment.

Reply:
@AlekseyTs Updated.